### ISSUE TITLE
Add a way to run bob with a specific list of updated resources

### ISSUE THREAD

**Issue by astrochili** [defold/defold#8490]
The build time with bob is longer than in Defold Editor.

This is noticeable when using the [Defold Kit](https://github.com/astrochili/vscode-defold) VS Code extension, which uses bob to build and run the game for debugging in VSCode. But in this case, when using Defold Kit most often you don't need to rebuild all the resources, only the modified ones.

Tracking of changed files is possible in VS Code, there is an API for that. So adding tracking to Defold Kit should not be difficult. 

This feature request is about adding some API for bob to rebuild only the list of changed resources and not waste time on unchanged ones.

Related issue — https://github.com/astrochili/vscode-defold/issues/33.

**Comment by swajj** [comment]
I would love to see bob incremental builds- I also build using bob, but mostly via shell scripting. I would really love if bob's output included a cache of file modification times or some other data so that I could run bob incrementally as long as that cache persisted. I maintain multiple concurrent build directories (client / server) so this would need to be cached by build. I see there's already a `_BobBuildState_` file in there- is this something that's already being tracked perhaps?

**Comment by ekharkunov** [comment]
Take a look at https://defold.com/manuals/caching-assets/. Maybe it will be helpful.

**Comment by britzl** [comment]
> Take a look at https://defold.com/manuals/caching-assets/. Maybe it will be helpful.

Yep, adding a local cache will surely help a bit. But I suppose one problem with bob compare to the editor is that the editor is loaded, ready and has the full project state in memory, which makes it really fast to rebuild changed assets. Bob on the other hand need to start a java vm, load the classes of bob.jar, load the entire project and the build it. I would like to point out that bob will not rebuild something that is already built, but it does ofc have to do some processing (read file and measure checksums) to figure out if it has to be rebuilt or not.

**Comment by astrochili** [comment]
Is this the equivalent of running the `build` command without the `distclean` command? I thought there was a cache in the `/build/` folder, which is reused when you run `bob build `again.

**Comment by astrochili** [comment]
May be we need I think we need measurements of some large project to see what takes the longest.

I don't have huge projects. My small ones with a dozen collections, several atlases and half a hundred lua scripts take 5 to 15 seconds to build second time, which is acceptable to me. But in the editor, yes, it is faster, takes about 2-3 seconds.

**Comment by JCash** [comment]
Bob already supports incremental builds. It does so by using a cache.

Are you sure you're not calling it with the `clean` command?
If doing so, it will rebuild all content.

Also note that bob was/is designed to be run once, not to be run as a server (like the editor), so you'll always pay for the startup time (compare with the loading of a project in the editor).

I'm not sure what we can do to speed up the initial loading of the project when starting bob.

**Comment by AGulev** [comment]
Also, the process is more complex than simply checking if a file has changed to decide on a rebuild. As Björn mentioned:

> but it does of course have to do some processing (read the file and measure checksums) to figure out if it has to be rebuilt or not.

However, it involves more than just reading the file and getting its checksum. We run the `create` stage for each builder (see `@Override public Task<Void> create(IResource input)` in each `*Builder.java`). This is crucial because the system does not operate on a `one input -> one output` basis. Some inputs produce multiple outputs, and some outputs require multiple inputs. Additionally, there are dependencies, where one output can be utilized as an input for another resource tree node.

The editor maintains the dependency tree in memory and always knows exactly what needs to be rebuilt at any point in time. Bob, on the other hand, cannot do this and must read this tree from the project files all the time (that's what the `create()` function in each builder does). That's what takes most of the time. You can check it yourself by producing html build report.

**Comment by astrochili** [comment]
> Are you sure you're not calling it with the clean command?

Sure, there is a difference. I don't have large projects, so maybe my measurements aren't representative, but I tried to record 4 logs and separate in them intervals that either take appreciable time or are very different between the first and second builds.

- [catsday_clean.log](https://github.com/defold/defold/files/14181845/catsday_clean.log) — a clean first build of [**catsday**](https://astronachos.com/catsday/).
- [catsday_second.log](https://github.com/defold/defold/files/14181844/catsday_second.log) — a second build of **catsday** without cleaning.
- [walker_clean.log](https://github.com/defold/defold/files/14181846/walker_clean.log) — a clean first build of [**walker**](https://github.com/astrochili/defold-kinematic-walker).
- [walker_second.log](https://github.com/defold/defold/files/14181847/walker_second.log) — a second build of **walker**.

Despite saving on matches (I'm personally happy with the current build speed anyway), what can we see here that's interesting?

- Generally **catsday** builds faster the second time (_1.3 sec vs. 4.2 sec `from 0% to 99%`_). But during the second build something happens between the last `99% to 100%` for 4.5 seconds, that doens't happen during the clean build, so the total time is the same — _7.4 vs 7.8 seconds_.
- The **walker** performs much better on the second build - 1.7 vs. 3.8 seconds. Again, the difference is in the process from `0% to 99%` - _0.6 vs. 2.2 seconds_. The interval from `99% to 100%` is also faster on the second build - _0.1 vs. 0.6 seconds_.

So, I don't know, if `0% to 99%` is what is cached and the rest is environment initialization, then there's really nothing to speed up.

**Comment by JCash** [comment]
Yeah, not sure what things are processed in the actual last bit (99% to 100%), and it seems worth investigating.

**Comment by AGulev** [comment]
> ### Total for **catsday** — **7.8 vs 7.4** seconds 🤷‍♂️.
> * `0% to 99%` — 4.1 vs 1.3 seconds ✅.
> * `99% to 100%` — 1.0 vs 4.5 seconds 🤷‍♂️.
> * `other`  — 2.7 vs 1.6 seconds ✅.
> 
> ### Total for **walker** — **3.8 vs 1.7** seconds ✅.
> * `0% to 99%` — 2.2 vs 0.6 seconds ✅.
> * `99% to 100%` — 0.6 vs 0.1 seconds ✅.
> * `other` — 1.0 vs 1.0 seconds ✅.


> Yeah, not sure what things are processed in the actual last bit (99% to 100%), and it seems worth investigating. So, same end goal, but a slightly different task.

It would be much easier to analyze what's going on if you provide time reports instead of Bob logs.

Initialization time isn't that long, really. The most time-consuming parts are unpacking and task creation. The flow looks like this:

1. In the current implementation, we create tasks for all the files in the project folder. Starting from version 1.9.3, task creation will be based on `game.project` and the project tree, depending on the dependencies in each file.
   
2. After task creation, we have enough information to know what exactly needs to be rebuilt based on the hash digest (which includes the SHA-1 of outputs and extra keys for some file types).
   
3. If the hash digest is the same, the build stage is skipped for those files.

4. Build stage.
   
5. Bundle stage, if needed.

If I had a time report (you can generate it using `--build-report`), I could tell you what exactly is taking so long in your case.

Also, it would be nice to compare it with the latest alpha, where the build process was rewritten.

-----

Technically, we can provide separate functionality (an argument for Bob) that will allow building individual files if needed and if we really want it.

However, it will be up to the developer to decide when and how to use it, as we cannot guarantee that everything will work correctly if not all dependencies in the tree above are built.

Off the top of my head, I don't foresee issues if this is used only with scripts, as scripts don't affect anything higher in the tree above the script. However, in the case of GameObjects and similar entities, it may result in unpredictable errors.

Also, after the refactoring done for version 1.9.3, it will not be rebuilding individual files anyway. Starting from 1.9.3, Bob will check all the dependencies of a file to determine if those dependencies should be rebuilt as well.

**Comment by AGulev** [comment]
I'm still not sure if we want such functionality. Using bob without `clean` or `distclean ` should robust now (in 1.10.2).

**Comment by AGulev** [comment]
@astrochili shared build reports and his command line:
[metrics.zip](https://github.com/user-attachments/files/20779129/metrics.zip)

```
$ /Applications/Defold.app/Contents/Resources/packages/jdk-21.0.5+11/bin/java" -cp "/Applications/Defold.app/Contents/Resources/packages/defold-5944eddf7392f6735768b5dfc495695b63ef3c76.jar" "com.dynamo.bob.Bob" --variant debug --output build/defoldkit build

[starting]
Working...
Reading classes...
2025-06-17 18:24:45 INFO
1% ...done!
Reading tasks...
Build Remote Engine...
Building engine...
2147483647% ...done!
Building...
Creating task builder with a fixed thread pool executor using 6 threads
Build tasks
Build tasks took 0.063000 s 
 99% ...done!
...done!
Engine build took 3.166000 s
100% ...done!
[exit 0]

$ cp '/Users/astro/Repo/gamedev/chestquest/build/arm64-osx/dmengine' to '/Users/astro/Repo/gamedev/chestquest/build/defoldkit/dmengine'
$ chmod +x "/Users/astro/Repo/gamedev/chestquest/build/defoldkit/dmengine"
$ /Users/astro/Repo/gamedev/chestquest/build/defoldkit/dmengine build/defoldkit/game.projectc
```

**Comment by AGulev** [comment]
The initial request may improve only this step, which takes 1.7s - not the biggest bottleneck, I'll check why engine build takes 3 sec when it should be almost immediately. 

![Image](https://github.com/user-attachments/assets/6a7dcf44-ab65-4851-999b-98bdba95a9fd)

**Comment by AGulev** [comment]
I'm closing this issue if favor of https://github.com/defold/defold/issues/10854
Because using of such functionality will make the build process fragile: Bob will not be able to track if files up in the tree should be rebuilt because of changes down in the tree.

Instead, we can optimize some of the stages to make build faster.
